gapic/src/main/com/google/gapid/models/Devices.java

```diff
@@ -295,4 +315,87 @@ public boolean isStadia() {
       return device.getConfiguration().getOS().getKind() == Device.OSKind.Stadia;
     }
   }
+
+  private static class DeviceValidationInfo {
+    private static DeviceValidationInfo instance = null;
+    private SingleInFlight rpcController = new SingleInFlight();
+    private final Client client;
+    private final Shell shell;
+    private Map<DeviceCaptureInfo, Boolean> status;
+
+    private DeviceValidationInfo(Client client, Shell shell) {
+      this.client = client;
+      this.shell = shell;
+      status = Maps.newHashMap();
+    }
+
+    private void validate(DeviceCaptureInfo device, Function<Boolean, Void> callback) {
+      if (device == null) {
+        return;
+      }
+      if (status.getOrDefault(device, null) != null) {
+        callback.apply(status.get(device));
+        return;
+      }
+
+      rpcController.start().listen(client.validateDevice(device.path),
+          new UiErrorCallback<Service.ValidateDeviceResponse, Service.ValidateDeviceResponse, Service.ValidateDeviceResponse>(shell, LOG) {
+
+        @Override
+        protected ResultOrError<Service.ValidateDeviceResponse, Service.ValidateDeviceResponse>
+          onRpcThread(Rpc.Result<Service.ValidateDeviceResponse> response) throws RpcException, ExecutionException {
+          try {
+            return success(response.get());
+          } catch (RpcException | ExecutionException e) {
+            throttleLogRpcError(LOG, "LoadData error", e);
+            return error(null);
+          }
+        }
+
+        @Override
+        protected void onUiThreadSuccess(Service.ValidateDeviceResponse response) {
+          updateValidationStatus(device, response);
+          callback.apply(getValidationStatus(device));
+        }
+
+        @Override
+        protected void onUiThreadError(Service.ValidateDeviceResponse response) {
+          LOG.log(WARNING, response.toString());
+          updateValidationStatus(device, response);
+          callback.apply(getValidationStatus(device));
+        }
+      });
+    }
+
+    protected static void updateValidationStatus(DeviceCaptureInfo device, Service.ValidateDeviceResponse response) {
+      if (instance == null) {
+        return;
+      }
+      if (response == null || response.hasError()) {
+        instance.status.put(device, false);
+        return;
+      }
+      instance.status.put(device, true);
+    }
+
+    public static void createInstance(Client client, Shell shell) {
+      instance = new DeviceValidationInfo(client, shell);
+    }
+
+    public static void validateDevice(DeviceCaptureInfo device, Function<Boolean, Void> callback) {
+      if (instance != null) {
+        instance.validate(device, callback);
+      }
+    }
+
+    public static void reset() {
```

Comment:
why? I don't think there is any reason to reset the validation results. In fact with caching we want it to persist between runs.

Reply:
Removed reset now. I had it as a way of not re-running validation, if it failed, unless the user presses the refresh button. Now, if the validation fails, just changing the device to something else and back will trigger it.